nntrainer/src/conv2d_layer.cpp

```diff
@@ -138,6 +176,90 @@ sharedConstTensor Conv2DLayer::backwarding(sharedConstTensor derivative,
     delBias.setZero();
   }
 
+  Tensor ret(input_dim.batch(), input_dim.channel(),
+             input_dim.height() + padding[0] * 2,
+             input_dim.width() + padding[1] * 2);
+  ret.setZero();
+
+#ifdef USE_BLAS
+  int status = ML_ERROR_NONE;
+  for (unsigned int b = 0; b < input_dim.batch(); ++b) {
+    std::vector<float> out(kernel_size[0] * kernel_size[1] *
+                           input_dim.channel() * filter_size);
+
+    Tensor inSub(
+      TensorDim(1, input_dim.channel(), input_dim.height(), input_dim.width()),
+      input.getAddress(b * input.getDim().getFeatureLen()));
+
+    status = conv2d_gemm(
+      derivative->getAddress(b * derivative->getDim().getFeatureLen()),
+      TensorDim(1, derivative->channel(), derivative->height(),
+                derivative->width()),
+      inSub,
+      TensorDim(1, 1, filter_size,
+                kernel_size[0] * kernel_size[1] * input_dim.channel()),
+      stride, padding, out.data(), out.size(), false);
+    if (status != ML_ERROR_NONE)
+      throw std::runtime_error("Backwarding Convolution failed.");
+
+    for (unsigned int i = 0; i < filter_size; ++i) {
+      Tensor &delK = paramsAt(i).grad;
+      Tensor &delBias = paramsAt(i + filter_size).grad;
+      float *del = delK.getData();
+      unsigned int s = kernel_size[0] * kernel_size[1] * input_dim.channel();
+
+      for (unsigned int k = 0; k < s; ++k) {
```

Comment:
I recommend you to tensor.sum_by_batch() for this. It uses blas internally if available and will be faster.

Reply:
out is not tensor. It just is vector. There might be copy and I'm not sure it is faster.

Reply:
you consider cblas_gemv here then (as this code is for blas only).

Reply:
You can directly use blas here as it is under USE_BLAS case. Then there will not be any copy.

Reply:
I will change with blas func. later.